domain/upgrade/state/state.go

```diff
@@ -0,0 +1,311 @@
+func NewState(factory domain.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+func (st *State) getUpgrade(ctx context.Context, tx *sqlair.TX, upgradeUUID string) (info, error) {
+	q := `
+SELECT * AS &info.* FROM upgrade_info`
+	s, err := sqlair.Prepare(q, info{})
+	if err != nil {
+		return info{}, errors.Annotatef(err, "preparing %q", q)
+	}
+
+	var upgradeInfo info
+	err = tx.Query(ctx, s).Get(&upgradeInfo)
+	if err != nil {
+		return info{}, errors.Trace(err)
+	}
+
+	return upgradeInfo, nil
+}
+
+func (st *State) getUpgradeNodes(ctx context.Context, tx *sqlair.TX, upgradeUUID string) ([]infoControllerNode, error) {
+	q := `
+SELECT * AS &infoControllerNode.*
+FROM   upgrade_info_controller_node
+WHERE  upgrade_info_controller_node.upgrade_info_uuid = $M.info_uuid`
+	s, err := sqlair.Prepare(q, infoControllerNode{}, sqlair.M{})
+	if err != nil {
+		return nil, errors.Annotatef(err, "preparing %q", q)
+	}
+
+	var nodeInfos []infoControllerNode
+	err = tx.Query(ctx, s, sqlair.M{"info_uuid": upgradeUUID}).GetAll(&nodeInfos)
+	if err != nil && err != sqlair.ErrNoRows {
+		return nil, errors.Trace(err)
+	}
+	return nodeInfos, nil
+
+}
+
+// CreateUpgrade creates an active upgrade to and from specified versions
+// and returns the upgrade's UUID
+func (st *State) CreateUpgrade(ctx context.Context, previousVersion, targetVersion version.Number) (string, error) {
+	db, err := st.DB()
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+
+	upgradeUUID := utils.MustNewUUID().String()
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		q := `
+INSERT INTO upgrade_info (uuid, previous_version, target_version, created_at) VALUES
+($M.uuid, $M.previous_version, $M.target_version, DATETIME('now'))`
+		s, err := sqlair.Prepare(q, sqlair.M{})
+		if err != nil {
+			return errors.Annotatef(err, "preparing %q", q)
+		}
+		return errors.Trace(tx.Query(ctx, s, sqlair.M{
+			"uuid":             upgradeUUID,
+			"previous_version": previousVersion.String(),
+			"target_version":   targetVersion.String(),
+		}).Run())
+	})
+
+	if err != nil {
+		return "", errors.Trace(err)
+	}
+	return upgradeUUID, nil
+}
+
+// SetControllerReady marks the supplied controllerID as being ready
+// to start a provided upgrade. All provisioned controllers need to
+// be ready before an upgrade can start
+// A controller node is ready for an upgrade if a row corresponding
+// to the controller is present in upgrade_info_controller_node
+func (st *State) SetControllerReady(ctx context.Context, upgradeUUID, controllerID string) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		nodeInfos, err := st.getUpgradeNodes(ctx, tx, upgradeUUID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for _, nodeinfo := range nodeInfos {
+			if nodeinfo.ControllerNodeID == controllerID {
+				return nil
+			}
+		}
```

Comment:
I think it would be better to push this in to the db layer and just do a select where any already have one, just return true.

```go
ok, err := st.alreadyUpgrading(ctx, tx, upgradeUUID, controllerID)
```